lib/api.rb

```diff
@@ -243,59 +314,159 @@ def validation_error(msg)
+     "Culture" => d.user.culture,
+     "TwoFactorEnabled" => d.user.totp_secret,
+     "Key" => d.user.key,
+     "PrivateKey" => d.user.private_key,
+     "SecurityStamp" => d.user.security_stamp,
+     "Organizations" => "[]",
+     "Object" => "profile",
+    }.to_json
+  end
+
+  # Used to update masterpassword
+  post "/accounts/password" do
+    content_type :json
+    response['access-control-allow-origin'] = '*'
+    d = device_from_bearer
+    if !d
+      return validation_error("invalid bearer")
     end
 
-    c = Cipher.new
-    c.user_uuid = d.user_uuid
-    c.update_from_params(params)
+   need_params(:key, :masterpasswordhash, :newmasterpasswordhash) do |p|
+     return validation_error("#{p} cannot be blank")
+   end
 
-    Cipher.transaction do
-      if !c.save
-        return validation_error("error saving")
-      end
+   if !params[:key].to_s.match(/^0\..+\|.+/)
+     return validation_error("Invalid key")
+   end
 
-      c.to_hash.merge({
-        "Edit" => true,
-      }).to_json
-    end
+   begin
+     Bitwarden::CipherString.parse(params[:key])
+   rescue Bitwarden::InvalidCipherString
+     return validation_error("Invalid key")
+   end
+
+   if d.user.password_hash == params[:masterpasswordhash]
+      d.user.key=params[:key]
+     d.user.password_hash=params[:newmasterpasswordhash]
+   else
+     return validation_error("Wrong current password")
+   end
+
+    User.transaction do
+     if !d.user.save
+       return validation_error("Unknown error")
+     end
+   end
+   ""
   end
 
-  # update a cipher
-  put "/ciphers/:uuid" do
+  # Used to update email
+  post "/accounts/email-token" do
+   content_type :json
+   response['access-control-allow-origin'] = '*'
+   validation_error("Not implemented yet")
+  end
+
+  #
+  # ciphers
+  #
+
+  # Import from keepass or others via web vault
+  post "/ciphers/import" do
+    content_type :json
+    response['access-control-allow-origin'] = '*'
     d = device_from_bearer
     if !d
       return validation_error("invalid bearer")
     end
 
-    c = nil
-    if params[:uuid].blank? ||
-    !(c = Cipher.find_by_user_uuid_and_uuid(d.user_uuid, params[:uuid]))
-      return validation_error("invalid cipher")
+    #First we create all the folders
+    params[:folders].each do |p|
+      f = Folder.new
+      f.user_uuid = d.user_uuid
+      f.update_from_params(p)
+      Folder.transaction do
+        if !f.save
+          return validation_error("error saving")
+        end
+      end
+    end
+
+    # We create each CipherString
+    i = 0
+    params[:ciphers].each do |p|
+      c = Cipher.new
+      c.user_uuid = d.user_uuid
+      c.update_from_params(p)
+      c.folder_uuid = Folder.find_by_user_uuid_and_name(d.user_uuid, params[:folders][params[:folderrelationships][i]["value"].to_i]["name"]).uuid
+      Cipher.transaction do
+        if !c.save
+          return validation_error("error saving")
+        end
+      end
+      i+=1
```

Comment:
This should just use `params[:ciphers].each_with_index do |p,i|` and avoid having to initialize `i` and increment it manually.